changes to compile on linux with gcc v12 + Dockerfile

CristiMacovei opened the pull request:
### Made changes to compile the project on linux targets using gcc version 12.2.

- a variable named `debug` was defined in two places in `dfu-util/`
	- changed `dfu-util/dfu.h`: removed declaration of `debug` variable
	- changed `dfu-util/sam7dfu.c`: added extern declaration of `debug`
	- kept definition in `dfu-util/main.c`

- a variable named `endianness` was already defined and redefined inside `ipsw-patch/main.c` and `xpwn/src/xpwn.cpp`
	- added `extern` modifier to the declaration

CristiMacovei committed:
Update code to build on Linux with GCC >= 12

Signed-off-by: Nicolae-Cristian Macovei <nicolae.macovei1412@stud.acs.upb.ro>

CristiMacovei committed:
Introduce Dockerfile

Signed-off-by: Nicolae-Cristian Macovei <nicolae.macovei1412@stud.acs.upb.ro>

razvand commented on Dockerfile:
```diff
@@ -0,0 +1,22 @@
+
+WORKDIR /
+
+ENTRYPOINT [ "/xpwn/builddir/ipsw-patch/xpwntool" ]
```
Add trailing newline.

razvand commented on Dockerfile:
```diff
@@ -0,0 +1,22 @@
+
+RUN apt update && apt upgrade -y
+
+RUN apt install -y git cmake build-essential libz-dev libssl-dev libbz2-dev libpng-dev libusb-dev
```
```suggestion
RUN set -xe; \
    apt-get -yqq update; \
    apt-get install -y git cmake build-essential libz-dev libssl-dev libbz2-dev libpng-dev libusb-dev \
    ;
```

razvand commented on Dockerfile:
```diff
@@ -0,0 +1,22 @@
+WORKDIR /xpwn
+
+RUN git checkout testing
+
```
This will eventually disappear when we merge this branch, right?

razvand reviewed (changes requested):
Please improve the commit messages. Use instructions [here](https://cbea.ms/git-commit/).

Also, I would add / update the `README` file with instructions on using the `Dockerfile`.

CristiMacovei committed:
Improve Dockerfile structure

This commit makes docker `COPY` from the local files, instead of pulling the codebase from git.

In addition, good-practice changes have been added to the `Dockerfile`, such as replacing `apt`
with `apt-get` and adding a newline to the end of the file.

Signed-off-by: Nicolae-Cristian Macovei <nicolae.macovei1412@stud.acs.upb.ro>

CristiMacovei committed:
Introduce usage instructions for Dockerfile

Signed-off-by: Nicolae-Cristian Macovei <nicolae.macovei1412@stud.acs.upb.ro>

CristiMacovei committed:
Introduce Docker pipeline

Signed-off-by: Nicolae-Cristian Macovei <nicolae.macovei1412@stud.acs.upb.ro>

razvand commented on Docker.md:
````diff
@@ -0,0 +1,27 @@
+You can also build your own image locally:
+
+```sh
+xpwn $ docker build -t <container_name> .
````
```suggestion
docker build -t <container_name> .
```

razvand commented on Docker.md:
```diff
@@ -0,0 +1,27 @@
+# Docker usage instructions
```
```suggestion
# Docker Usage Instructions
```

razvand commented on Docker.md:
```diff
@@ -0,0 +1,27 @@
+
+### Setup
+
+Container images are available on [the github container registry](https://github.com/orgs/malus-security/packages/container/package/xpwn) (and soon on the gitlab container registry).
```
```suggestion
Container images are available on [the GitHub Container Registry](https://github.com/orgs/malus-security/packages/container/package/xpwn) (and soon on the GitLab Container Registry).
```

razvand reviewed (changes requested):
Update your commit messages to follow [the best practices in creating commits and commit descriptions](https://cbea.ms/git-commit/).

That is:

1. Use your full name for signing-off and for commit authorship: `Cristian Macovei <nicolae.macovei1412@stud.acs.upb.ro>`.
1. Start sentences with capital letters, e.g. use `Improve` instead of `improve`.
1. When you add a new file, do not use the verb `Add`. Use the verb `Introduce`.
1. Use capitalized words for names, e.g. `Git` instead of `git`, `GCC` instead of `GCC`.
1. Add a detailed description for commit messages, e.g. the commit `add Docker pipeline` (the first line should be `Introduce Docker pipeline` - followed by an extended description). Same for `add Dockerfile` and `changes to compile on linux with gcc v12` (which should be `Update code to build on Linux with GCC >= 12`.
